rp2040-hal/src/pio.rs

```diff
@@ -0,0 +1,911 @@
+        sm.set_clock_divisor(self.clock_divisor);
+
+        sm.sm().sm_execctrl.write(|w| {
+            w.side_en().bit(self.side_set.optional());
+            w.side_pindir().bit(self.side_set.pindirs());
+
+            unsafe {
+                w.jmp_pin().bits(self.jmp_pin);
+            }
+
+            if let Some(inline_out) = self.inline_out {
+                w.inline_out_en().bit(true);
+                unsafe {
+                    w.out_en_sel().bits(inline_out);
+                }
+            } else {
+                w.inline_out_en().bit(false);
+            }
+
+            w.out_sticky().bit(self.out_sticky);
+
+            unsafe {
+                w.wrap_top().bits(offset as u8 + self.wrap.source);
+                w.wrap_bottom().bits(offset as u8 + self.wrap.target);
+            }
+
+            let n = match self.mov_status {
+                MovStatusConfig::Tx(n) => {
+                    w.status_sel().bit(false);
+                    n
+                }
+                MovStatusConfig::Rx(n) => {
+                    w.status_sel().bit(true);
+                    n
+                }
+            };
+            unsafe {
+                w.status_n().bits(n);
+            }
+
+            w
+        });
+
+        sm.sm().sm_shiftctrl.write(|w| {
+            let (fjoin_rx, fjoin_tx) = match self.fifo_join {
+                Buffers::RxTx => (false, false),
+                Buffers::OnlyTx => (false, true),
+                Buffers::OnlyRx => (true, false),
+            };
+            w.fjoin_rx().bit(fjoin_rx);
+            w.fjoin_tx().bit(fjoin_tx);
+
+            unsafe {
+                // TODO: Encode 32 as zero, and error on 0
+                w.pull_thresh().bits(self.pull_threshold);
+                w.push_thresh().bits(self.push_threshold);
+            }
+
+            w.out_shiftdir().bit(self.out_shiftdir.bit());
+            w.in_shiftdir().bit(self.in_shiftdir.bit());
+
+            w.autopull().bit(self.autopull);
+            w.autopush().bit(self.autopush);
+
+            w
+        });
+
+        sm.sm().sm_pinctrl.write(|w| {
+            unsafe {
+                w.sideset_count().bits(self.side_set.bits());
+                w.set_count().bits(self.set_count);
+                w.out_count().bits(self.out_count);
+
+                w.in_base().bits(self.in_base);
+                w.sideset_base().bits(self.side_set_base);
+                w.set_base().bits(self.set_base);
+                w.out_base().bits(self.out_base);
+            }
+
+            w
+        });
+
+        // Restart SM and its clock
+        sm.restart();
+        sm.reset_clock();
+
+        // Set starting location by setting the state machine to execute a jmp
+        // to the beginning of the program we loaded in.
+        #[allow(clippy::unusual_byte_groupings)]
+        let mut instr = 0b000_00000_000_00000; // JMP 0
+        instr |= offset as u16;
+        sm.set_instruction(instr);
```

Comment:
This looks less magical to me:
```suggestion
let instr = pio::InstructionOperands::JMP {
    condition: pio::JmpCondition::Always,
    address: offset as u8
};
sm.set_instruction(instr.encode());
```